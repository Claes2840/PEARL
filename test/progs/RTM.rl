init: entry
      S += BLANK // BLANK and SLASH should be defined in input
      if pc_max = 0 goto stop else loop

loop: fi Q1[pc] = Start from init else act4
      if Q1[pc] = End goto stop else act1

stop: fi pc_max = 0 from init else loop
      exit

act1: from loop 
      if Q = Q1[pc] && S = S1[pc] goto write else act2

write: from act1
       Q += Q2[pc] - Q1[pc]
       S += S2[pc] - S1[pc]
       goto act2

act2: fi Q = Q2[pc] && S = S2[pc] from write else act1
      if Q = Q1[pc] && S1[pc] = SLASH goto move else act3

act3: fi Q = Q2[pc] && S1[pc] = SLASH from move1 else act2
      pc += 1
      if pc = pc_max goto clear else act4

clear: from act3
       pc -= pc_max
       goto act4

act4: fi pc = 0 from clear else act3
      goto loop

move: from act2
      Q += Q2[pc] - Q1[pc]
      if S2[pc] = LEFT goto left else right

left: from move
      if empty S_right && S = BLANK goto left_1b else left_1p

left_1b: from left // MERGE? 1
         S -= BLANK
         goto left1

left_1p: from left
         push S S_right
         goto left1

left1: fi empty S_right from left_1b else left_1p
       if empty S_left goto left_2b else left_2p

left_2b: from left1 // MERGE? 2
         S += BLANK
         goto left2

left_2p: from left1
         pop S S_left
         goto left2

left2: fi empty S_left && S = BLANK from left_2b else left_2p
       goto move1

right: from move
       if empty S_left && S = BLANK goto right_1b else right_1p

right_1b: from right // MERGE? 1
          S -= BLANK
          goto right1

right_1p: from right
          push S S_left
          goto right1

right1: fi empty S_left from right_1b else right_1p
        if empty S_right goto right_2b else right_2p

right_2b: from right1 // MERGE? 2
          S += BLANK
          goto right2

right_2p: from right1
          pop S S_right
          goto right2

right2: fi empty S_right && S = BLANK from right_2b else right_2p
        goto move1

move1: fi S2[pc] = LEFT from left2 else right2
       goto act3
