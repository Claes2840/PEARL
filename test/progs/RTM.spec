// Symbols and required values
BLANK = 32
SLASH = 47
LEFT = 60
RIGHT = 62
S = 0

// Starting state, both numbers must be equal
Q = 1
Start = 1

// End state and state machine definition
End = 6
pc = 0
pc_max = 8
Q1 = [ 1,  2, 3, 3,  3,  4, 5,  5]
S1 = [32, 47, 0, 1, 32, 47, 0, 32]
S2 = [32, 62, 1, 0, 32, 60, 0, 32]
Q2 = [ 2,  3, 4, 2,  4,  5, 4,  6]

// Q = 2, pc = 1