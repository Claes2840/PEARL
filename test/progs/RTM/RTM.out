(S_left S S_right) -> (S_left S S_right)

init_9:
	entry
	assert((S = 'BLANK))
	S ^= 'BLANK
	S ^= 'BLANK
	goto act2_1

act2_1:
	fi (S = 'BLANK)
		from init_9
		else write_3
	assert(!((S = 'SLASH)))
	if ((S_left = 'nil) && (S = 'BLANK))
		goto right_1b_2
		else right_1p_2

right_1b_2:
	from act2_1
	S ^= 'BLANK
	goto right1_2

right1_2:
	fi (S_left = 'nil)
		from right_1b_2
		else right_1p_2
	if (S_right = 'nil)
		goto right_2b_2
		else right_2p_2

right_2b_2:
	from right1_2
	S += 'BLANK
	goto right2_2

right2_2:
	fi ((S_right = 'nil) && (S = 'BLANK))
		from right_2b_2
		else right_2p_2
	if (S = '0)
		goto write_4
		else act2_4

write_4:
	from right2_2
	S ^= '0
	S ^= '1
	goto act2_5

act2_5:
	fi (S = '1)
		from write_4
		else write_8
	goto act2_6

act2_6:
	fi (S = 'BLANK)
		from act2_3
		else act2_5
	assert(!((S = 'SLASH)))
	if ((S_right = 'nil) && (S = 'BLANK))
		goto left_1b_7
		else left_1p_7

left_1b_7:
	from act2_6
	S ^= 'BLANK
	goto left1_7

left1_7:
	fi (S_right = 'nil)
		from left_1b_7
		else left_1p_7
	if (S_left = 'nil)
		goto left_2b_7
		else left_2p_7

left_2b_7:
	from left1_7
	S ^= 'BLANK
	goto left2_7

left2_7:
	fi ((S_left = 'nil) && (S = 'BLANK))
		from left_2b_7
		else left_2p_7
	if (S = '0)
		goto write_8
		else act2_8

write_8:
	from left2_7
	S ^= '0
	S ^= '0
	assert((S = '0))
	goto act2_5

act2_8:
	from left2_7
	assert((S = 'BLANK))
	S ^= 'BLANK
	S ^= 'BLANK
	assert((S = 'BLANK))
	exit

left_2p_7:
	from left1_7
	(S . S_left) <- S_left
	goto left2_7

left_1p_7:
	from act2_6
	S_right <- (S . S_right)
	goto left1_7

act2_4:
	from right2_2
	if (S = '1)
		goto write_3
		else act2_3

write_3:
	from act2_4
	S ^= '1
	S ^= '0
	assert((S = '0))
	goto act2_1

act2_3:
	from act2_4
	assert((S = 'BLANK))
	S ^= 'BLANK
	S ^= 'BLANK
	goto act2_6

right_2p_2:
	from right1_2
	(S . S_right) <- S_right
	goto right2_2

right_1p_2:
	from act2_1
	S_left <- (S . S_left)
	goto right1_2
